### PR TITLE
DDF-4368 Updating search and result forms without page refresh

### PR DESCRIPTION
**Pull request by hayleynorton**
#### What does this PR do?
Allows you to see new search and result forms in the grid without refreshing the page
#### Who is reviewing it? 
@Bdthomson @andrewzimmer 
#### Ask 2 committers to review/merge the PR and tag them here.
@andrewkfiedler
@bdeining
@djblue
#### How should this be tested?
- Search Forms
  - Navigate to the Search Forms page
  - Create a new form
  - Verify that a success message appears saying the form was successfully saved
  - Ensure that you can see the new form in the grid without refreshing the page
  - Click the interactions drop-down
  - Verify that you see _all_ of the following options:
    - Make Default Form
    - Share Form
    - Edit Form
    - Delete Form
  - Make sure you can successfully complete all these actions
  - When editing, change the title and search term
  - After editing and saving, ensure that you see a success message saying the form was successfully saved
  - Refresh the page and verify that all your edits persisted
- Result Forms
  - Navigate to the Result Forms page
  - Create a new form
  - Verify that success message appears saying the form was successfully saved
  - Ensure that you can see the new form in the grid without refreshing the page
  - Click the interactions drop-down
  - Verify that you see _only_ the following options:
    - Share Form
    - Delete Form
  - Make sure you can successfully complete all these actions
  - When editing, change all three fields
  - After editing and saving, ensure that you see a success message saying the form was successfully saved
  - Refresh the page and verify that all your edits persisted

Additionally, try to edit forms (PUT request) with the target metacard ID in the HTTP JSON body different than the ID in the query param, and ensure the request fails
#### What are the relevant tickets?
[DDF-4368](https://codice.atlassian.net/browse/DDF-4368)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 1]
```diff
@@ -173,6 +174,23 @@ public void init() {
         GSON::toJson);
 
     put(
+        "/forms/query/:id",
```
Comment (Bdthomson) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: It's a little awkward that we are taking in the id and then not using it. Maybe we can refactor the `doCreateOrUpdate` function call to two separate functions where one takes in an id.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 2]
```diff
@@ -189,6 +207,23 @@ public void init() {
                             .orElse(null))),
         GSON::toJson);
 
+    put(
+        "/forms/result/:id",
```
Comment (Bdthomson) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: Same as above.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 3]
```diff
@@ -311,8 +346,13 @@ private String safeGetBody(Request req) {
...
     }
-    catalogFramework.create(new CreateRequestImpl(metacard));
-    return ImmutableMap.of(RESP_MSG, "Successfully created");
+    Metacard createdMetacard =
+        catalogFramework.create(new CreateRequestImpl(metacard)).getCreatedMetacards().get(0);
+    return ImmutableMap.of(
+        "id",
+        createdMetacard.getAttribute("id").getValue(),
+        "createdOn",
+        createdMetacard.getAttribute("created").getValue());
```
Comment (Bdthomson) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: ImmutableMap's builder is a bit cleaner to read.

```java
    return ImmutableMap.<String, String>builder()
            .put("id", createdMetacard.getAttribute("id").getValue())
            .put("created", "created")
            .build();
```

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-advanced.view.js [hunk 4]
```diff
@@ -66,7 +66,9 @@ module.exports = Marionette.LayoutView.extend({
       this.model.get('filterTree') !== undefined
     ) {
       this.queryAdvanced.currentView.deserialize(
-        cql.simplify(JSON.parse(this.model.get('filterTree')))
+        typeof this.model.get('filterTree') === 'string'
+          ? JSON.parse(this.model.get('filterTree'))
+          : this.model.get('filterTree')
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-advanced.view.js]: What does this do?

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.view.js [hunk 5]
```diff
@@ -88,11 +88,10 @@ module.exports = Marionette.LayoutView.extend({
     )
   },
   setupTitleInput: function() {
-    let currentValue = this.model.get('name') ? this.model.get('name') : ''
     this.basicTitle.show(
       new PropertyView({
         model: new Property({
-          value: [currentValue],
+          value: [this.model.get('title')],
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.view.js]: Is this guaranteed to be here? You might want to just update this deleted line to use `title` instead.
```java
    let currentValue = this.model.get('title') ? this.model.get('title') : ''
```

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.hbs [hunk 6]
```diff
@@ -52,16 +52,16 @@
...
                     {{createdOn}}
             </span>
         <span class="search-form-contents">
             <span class="fa fa-cloud"></span>
             {{createdBy}}
         </span>
-        <span class="choice-actions is-button" title="Shows a list of actions to take on the search forms" 
-        data-help="Shows a list of actions to take on the search forms.">
+        <span class="choice-actions is-button" title="Shows a list of actions to take on the result forms"
+        data-help="Shows a list of actions to take on the result forms.">
```
Comment (Bdthomson) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.hbs]: Was this an accident?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 7]
```diff
@@ -311,8 +346,13 @@ private String safeGetBody(Request req) {
         return ImmutableMap.of(RESP_MSG, "Successfully updated");
       }
     }
-    catalogFramework.create(new CreateRequestImpl(metacard));
-    return ImmutableMap.of(RESP_MSG, "Successfully created");
+    Metacard createdMetacard =
+        catalogFramework.create(new CreateRequestImpl(metacard)).getCreatedMetacards().get(0);
+    return ImmutableMap.of(
+        "id",
+        createdMetacard.getAttribute("id").getValue(),
```
Comment (willwill96) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: ✏️ `Core.ID`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java (hunk 3 above)
Comment (willwill96) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: ✏️ `Core.CREATED`

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-advanced.view.js [hunk 8]
```diff
@@ -66,7 +66,9 @@ module.exports = Marionette.LayoutView.extend({
       this.model.get('filterTree') !== undefined
     ) {
       this.queryAdvanced.currentView.deserialize(
-        cql.simplify(JSON.parse(this.model.get('filterTree')))
+        typeof this.model.get('filterTree') === 'string'
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-advanced.view.js]: ❓ Why is `cql.simplify` being removed?

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.view.js (hunk 5 above)
Comment (hayleynorton) [reply ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.view.js]: Every form gets the default defined in `search-form.js`, which I changed to `''`, so title is guaranteed to be at least the empty string

File: ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-advanced.view.js (hunk 4 above)
Comment (hayleynorton) [reply ui/packages/catalog-ui-search/src/main/webapp/component/query-advanced/query-advanced.view.js]: Sometimes `this.model.get('filterTree')` is already a JSON, which causes issues when you try to turn it into a JSON again. But sometimes it's a string, which needs to be converted to a JSON. This ternary avoids issues stemming from those inconsistencies

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form-editor/search-form-editor.view.js [hunk 9]
```diff
@@ -225,4 +193,18 @@ module.exports = Marionette.LayoutView.extend({
       },
     })
   },
+  successMessage: function() {
+    announcement.announce({
+      title: 'Success',
+      message: 'Result form successfully saved',
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-form-editor/search-form-editor.view.js]: ❓ Should this say `Search form`?

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form-editor/search-form-editor.view.js [hunk 10]
```diff
@@ -225,4 +193,18 @@ module.exports = Marionette.LayoutView.extend({
...
+      type: 'success',
+    })
+  },
+  errorMessage: function() {
+    announcement.announce({
+      title: 'Error',
+      message: 'Result form failed to save',
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-form-editor/search-form-editor.view.js]: ❓ Should this say `Search form`?

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.view.js (hunk 5 above)
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.view.js]: ❓ Why is the null check being removed?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 11]
```diff
@@ -311,8 +346,13 @@ private String safeGetBody(Request req) {
...
+        "createdOn",
```
Comment (willwill96) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: ✏️ Could create a constant

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java (hunk 7 above)
Comment (hayleynorton) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: Oh yeah, thanks for the reminder! I meant to go back and do that but forgot

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form-editor/search-form-editor.view.js (hunk 9 above)
Comment (hayleynorton) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-form-editor/search-form-editor.view.js]: Yikes, yes. Thank you

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.view.js [hunk 12]
```diff
@@ -180,55 +179,33 @@ module.exports = Marionette.LayoutView.extend({
...
-          owner: _this.model.get('owner'),
-          descriptors: _this.model.get('descriptors'),
-          description: _this.model.get('description'),
-          accessGroups: _this.model.get('accessGroups'),
-          accessIndividuals: _this.model.get('accessIndividual'),
-          accessAdministrators: _this.model.get('accessAdministrators'),
-        })
-        ResultFormCollection.getResultCollection().toggleUpdate()
-        _this.cleanup()
+    var collection = ResultFormsCollection.getCollection()
```
Comment (andrewkfiedler) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-form.view.js]: Could be const

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 13]
```diff
@@ -311,8 +349,12 @@ private String safeGetBody(Request req) {
...
       }
     }
-    catalogFramework.create(new CreateRequestImpl(metacard));
-    return ImmutableMap.of(RESP_MSG, "Successfully created");
+    Metacard createdMetacard =
+        catalogFramework.create(new CreateRequestImpl(metacard)).getCreatedMetacards().get(0);
+    return ImmutableMap.<String, Object>builder()
+        .put(Core.ID, createdMetacard.getId())
+        .put(CREATED_ON, createdMetacard.getAttribute(Core.CREATED).getValue())
+        .build();
```
Comment (Lambeaux) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: ❗️ Now that `PUT` is no longer overloaded to perform the tasks of both create and update, we should enforce proper HTTP semantics and split this method into two. Each can validate things a bit more harshly and fail fast if the request is not well formed, per our expectations. This mostly involves when a caller can provide an ID and when they cannot.

**Comment by Lambeaux** [comment]
Make a note that during the hero, try `PUT` requests with the target metacard ID in the HTTP JSON body *different* than the ID in the query param, and ensure the request fails.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 14]
```diff
@@ -286,35 +329,69 @@ private String safeGetBody(Request req) {
...
+  }
+
+  private Map<String, Object> doUpdate(
+      Response response, @Nullable Metacard metacard, @Nullable String expectedId)
       throws IngestException, SourceUnavailableException, FederationException,
           UnsupportedQueryException {
     if (metacard == null) {
+      LOGGER.debug("Form/Result metacard was null");
```
Comment (Bdthomson) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: Might want to add the metacard id to this log statement.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java [hunk 15]
```diff
@@ -286,35 +329,69 @@ private String safeGetBody(Request req) {
...
+      throws IngestException, SourceUnavailableException {
+    if (metacard == null) {
+      response.status(400);
+      return ImmutableMap.of(RESP_MSG, "Could not create metacard, no valid template specified");
+    }
+    if (metacard.getId() != null) {
+      LOGGER.debug(
+          "Cannot specify an ID [{}] when creating form/result metacard", metacard.getId());
+      response.status(400);
+      return ImmutableMap.of(RESP_MSG, "Could not create metacard, ID attribute not allowed");
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: `ID attribute not allowed` is misleading; is that actually what is happening in the code?

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-forms.view.js [hunk 16]
```diff
@@ -43,9 +43,7 @@ module.exports = SearchFormViews.extend({
       return
     }
 
-    lightboxInstance.model.updateTitle(
-      this.model.get('type') === 'new-result' ? '' : this.model.get('name')
-    )
+    lightboxInstance.model.updateTitle(this.model.get('title'))
```
Comment (bdeining) [inline ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-forms.view.js]: we no longer need the type check here?

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.hbs (hunk 6 above)
Comment (bdeining) [inline ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.hbs]: why was this changed?  is this not the handle bars for search forms?

File: ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-forms.view.js (hunk 16 above)
Comment (hayleynorton) [reply ui/packages/catalog-ui-search/src/main/webapp/component/result-form/result-forms.view.js]: The default title for a new result form, as defined in `search-form.js`, is `''`, so this check isn't necessary

File: ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.hbs (hunk 6 above)
Comment (hayleynorton) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.hbs]: Yes, but search and result forms share a lot of the same files, including this one. On line 53, there's a tag for `#is type result`, so the code within that tag is only used for result forms

Comment (bdeining) [reply ui/packages/catalog-ui-search/src/main/webapp/component/search-form/search-form.hbs]: 👍

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java (hunk 15 above)
Comment (Lambeaux) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/SearchFormsApplication.java]: When creating a search form, it's important to note that the framework should be responsible for assigning the ID. If someone tries to include it in the input, we fail, because they are likely expecting a metacard with **that** ID to be created after the request when in reality it got a random ID from the framework. 

We could be more generic about it. Or we could use something similar to the log message above. I'm good with either.

**Comment by hayleynorton** [comment]
build now

**Comment by gjvera** [comment]
hero build in progress

**Comment by hayleynorton** [comment]
build now

**Comment by gjvera** [comment]
🎉 Hero Successful 🎉 
Created Search Forms and Result forms. 
All new forms were shown instantly 
Changed the title and content for search forms and result forms and refreshed. All changes were persisted
Attempted a PUT request with a bad query ID and the request failed

LGTM!!
